[Multilane] Segments with multiple lanes

agalbachicar opened the pull request:
Towards #4934 completion, this PR makes `multilane::Segment`s able to create and hold more than one `multilane::Lane`.

Previous PR ( #6985  ) changed `multilane::RoadCurve` interface to provide support of lateral displacements of `multilane::Lane`s, which is used by this code to separate `multilane::Lane`'s centerlines from the reference curve of their parent `multilane::Segment` and distribute them along the _driveable bounds_.

Some new attributes are introduced to `multilane::Segment` class which are necessary to create equally spaced `multilane::Lane`:

- `num_lanes`: the number of lanes that segment will be allowed to build.
- `r_spacing`: the constant lateral distance between lane's centerlines.
- `r0`: this attributed was introduced in the last PR, and now is promoted to the `multilane::Segment`. It means the lateral distance from the reference curve to the first lane centerline.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7127)
<!-- Reviewable:end -->

agalbachicar commented:
+@maddog-tri for feature review please.

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-:-Kv8PNsKsAcRGjRbfgYf:b-uocvrj)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
First pass done (basically suggesting going back and making this interface much simpler).

---

Reviewed 2 of 9 files at r1.
Review status: 2 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/multilane/BUILD, line 146 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvApyELX6J3ePq9UAUh:-KvApyELX6J3ePq9UAUi:b34l58t) ([raw file](https://github.com/robotlocomotion/drake/blob/44f923541c13ef5e5ecd5bafa2d50a22969a8d5e/drake/automotive/maliput/multilane/BUILD#L146)):*
> ```Python
>     deps = [
>         ":lanes",
>         "//drake/automotive/maliput/api:maliput_types_compare",
> ```

(This is now in `.../maliput/api/test_utilities:maliput_types_compare`.)

---

*[drake/automotive/maliput/multilane/segment.cc, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvAuzU_vEPZwzMcAPQU:-KvAuzU_vEPZwzMcAPQV:b-hwnux8) ([raw file](https://github.com/robotlocomotion/drake/blob/44f923541c13ef5e5ecd5bafa2d50a22969a8d5e/drake/automotive/maliput/multilane/segment.cc#L31)):*
> ```cpp
>   // Ensures that all lanes are set inside Segment bounds.
>   const double last_lane_centerline =
>       r0 + static_cast<double>(num_lanes_ - 1) * std::copysign(r_spacing_, r0_);
> ```

Why the `copysign()`?  Lane indices are required to increase to the left, in the `+r` direction.

---

*[drake/automotive/maliput/multilane/segment.h, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvAvwtve3R4OtIzC8dO:-KvAvwtw8g2FnSQ5tE0n:b-12h4yf) ([raw file](https://github.com/robotlocomotion/drake/blob/44f923541c13ef5e5ecd5bafa2d50a22969a8d5e/drake/automotive/maliput/multilane/segment.h#L55)):*
> ```objc
>   Segment(const api::SegmentId& id, api::Junction* junction,
>           std::unique_ptr<RoadCurve> road_curve, int num_lanes, double r0,
>           double r_spacing, double r_min, double r_max,
> ```

This is a weird interface:  it's tightly controlling the offsets of each Lane, but the caller still has to count calls to `NewLane()` to make sure the id and bounds are associated with the correct Lane index.  My recommendation:  take all the offset/spacing parameters out of `Segment()`, make them freely settable via `NewLane()`, but assert that the various constraints are met by the caller.  It will be much simpler --- almost no code changes from the original:

  * `Segment()` goes back to its original arguments (losing `num_lanes`, `r0`, and `r_spacing`).
    * `Segment()` only needs to replace the `lane_` member with `lanes_`.
  * `NewLane()` goes back to its original arguments (regaining `r0`).
    * `NewLane()` gets an additional assertion that its `r0` is strictly greater than the `r0` of the last Lane on `lanes_` (if any).

(In other words:  if you're going to allow lane_bounds to be anything for each Lane, then you might as well allow r0 to be anything, too.  The `Builder`, on the other hand, will be even more simple and thus more restrictive, so it will enforce a relationship between bounds, offset, and constant spacing.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-:-KvB0FvjL0Zy1FAliwd8:bqmcx4u)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-xenial-gcc-experimental-matlab please
@drake-jenkins-bot linux-xenial-clang-experimental-matlab please

agalbachicar commented:
I've addressed all the comments. PTAL.

---

Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/multilane/BUILD, line 146 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvApyELX6J3ePq9UAUh:-KvDLGV4PGp5aLapwpm4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/44f923541c13ef5e5ecd5bafa2d50a22969a8d5e/drake/automotive/maliput/multilane/BUILD#L146)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

(This is now in `.../maliput/api/test_utilities:maliput_types_compare`.)
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/segment.cc, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvAuzU_vEPZwzMcAPQU:-KvDLNwJt0GpKShJ8c4Z:bpns611) ([raw file](https://github.com/robotlocomotion/drake/blob/44f923541c13ef5e5ecd5bafa2d50a22969a8d5e/drake/automotive/maliput/multilane/segment.cc#L31)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Why the `copysign()`?  Lane indices are required to increase to the left, in the `+r` direction.
</blockquote></details>

Done. Constructor code was restored to the header since there is no change to it.

---

*[drake/automotive/maliput/multilane/segment.h, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvAvwtve3R4OtIzC8dO:-KvDLqTyJ9lYb3V6tf12:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/44f923541c13ef5e5ecd5bafa2d50a22969a8d5e/drake/automotive/maliput/multilane/segment.h#L55)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

This is a weird interface:  it's tightly controlling the offsets of each Lane, but the caller still has to count calls to `NewLane()` to make sure the id and bounds are associated with the correct Lane index.  My recommendation:  take all the offset/spacing parameters out of `Segment()`, make them freely settable via `NewLane()`, but assert that the various constraints are met by the caller.  It will be much simpler --- almost no code changes from the original:

  * `Segment()` goes back to its original arguments (losing `num_lanes`, `r0`, and `r_spacing`).
    * `Segment()` only needs to replace the `lane_` member with `lanes_`.
  * `NewLane()` goes back to its original arguments (regaining `r0`).
    * `NewLane()` gets an additional assertion that its `r0` is strictly greater than the `r0` of the last Lane on `lanes_` (if any).

(In other words:  if you're going to allow lane_bounds to be anything for each Lane, then you might as well allow r0 to be anything, too.  The `Builder`, on the other hand, will be even more simple and thus more restrictive, so it will enforce a relationship between bounds, offset, and constant spacing.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-:-KvDPlqvvI3hDaK9gXrV:b76ci8)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> !!

---

Reviewed 9 of 10 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-:-KvEKHr6KdfmV1aUNhIr:bks4l3j)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
+@ericcousineau-tri for platform review please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-:-KvS4HXGF4KkytX_wfBl:b72l57d)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some minor BTWs

(Also, some of the pointers in testing code can be made `T* const` / `const T* const`, but I won't block on this.)

---

Reviewed 6 of 10 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/multilane/segment.h, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvTJRH8aQZsiKWNNRmY:-KvTJRH8aQZsiKWNNRmZ:b-tgoqry) ([raw file](https://github.com/robotlocomotion/drake/blob/6eea1c0791ca3819af56f404fd9434dcbfd6c476/drake/automotive/maliput/multilane/segment.h#L61)):*
> ```objc
>   ///
>   /// Driveable bounds of the lane will be derived based on the lateral offset
>   /// of it so as to reach r_min and r_max distances (see class constructor
> ```

For Doxygen parameters of length > 1, can you use backticks? ``

---

*[drake/automotive/maliput/multilane/test/multilane_lanes_test.cc, line 773 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvTKEeNo4JDnoR5B-da:-KvTKEeNo4JDnoR5B-db:b-fhmyeo) ([raw file](https://github.com/robotlocomotion/drake/blob/6eea1c0791ca3819af56f404fd9434dcbfd6c476/drake/automotive/maliput/multilane/test/multilane_lanes_test.cc#L773)):*
> ```cpp
>   // surroundings of each Lane. No saturation is tested.
>   lane_spacing = 0.;
>   api::GeoPosition nearest_position;
> ```

BTW For `nearest_position` and `distance`, if you do not need to re-use these and it does not negatively impact performance badly, can you narrow their scope to the innermost loop?

---

*[drake/automotive/maliput/multilane/test/multilane_lanes_test.cc, line 925 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvTPx0ME-D06tJE_CQW:-KvTPx0NluUJB-FHaMdx:blaufc1) ([raw file](https://github.com/robotlocomotion/drake/blob/6eea1c0791ca3819af56f404fd9434dcbfd6c476/drake/automotive/maliput/multilane/test/multilane_lanes_test.cc#L925)):*
> ```cpp
>         const double effective_angle = p * kDTheta + kTheta0;
>         const api::GeoPosition geo_point = api::GeoPosition::FromXyz(
>             kGeoCenter +
> ```

BTW Does this look like `clang-format`s output?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-:-KvTR5SzUZ2yIRlD427l:b-53640p)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
I have addressed the comments. PTAL.

Regarding the `T* const` / `const T* const` I guess that a code-refactor PR would be better. What do you think?

---

Review status: 7 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/multilane/segment.h, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvTJRH8aQZsiKWNNRmY:-KvTb4F4GOgBtKV1gE76:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6eea1c0791ca3819af56f404fd9434dcbfd6c476/drake/automotive/maliput/multilane/segment.h#L61)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

For Doxygen parameters of length > 1, can you use backticks? ``
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_lanes_test.cc, line 773 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvTKEeNo4JDnoR5B-da:-KvTbS_H_6ml2280v3n1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6eea1c0791ca3819af56f404fd9434dcbfd6c476/drake/automotive/maliput/multilane/test/multilane_lanes_test.cc#L773)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW For `nearest_position` and `distance`, if you do not need to re-use these and it does not negatively impact performance badly, can you narrow their scope to the innermost loop?
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/test/multilane_lanes_test.cc, line 925 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-KvTPx0ME-D06tJE_CQW:-KvTcslMldi7IGV9EAqS:brinils) ([raw file](https://github.com/robotlocomotion/drake/blob/6eea1c0791ca3819af56f404fd9434dcbfd6c476/drake/automotive/maliput/multilane/test/multilane_lanes_test.cc#L925)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Does this look like `clang-format`s output?
</blockquote></details>

Yes. I re-ran clang-format on this snippet and I didn't get any difference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-:-KvTifiVooNHFF8Mdg8Y:b9tq72y)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Looks great, and that sounds good to me. Thanks!

---

Reviewed 2 of 10 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7127#-:-KvTkcQwOugTtWwXUQcm:b-ifen00)*
<!-- Sent from Reviewable.io -->
